enhancements/network/ingress-node/ingress-node-firewall-with-ebpf.md

```diff
@@ -0,0 +1,284 @@
+|  0.0.0.0/0  |   TCP    |  22  |  SSH    |
+|  0.0.0.0/0  |   UDP    |  68  | DHCP    |
+|  0.0.0.0/0  |   TCP    | 6443 | Kubernetes API server access |
+|  0.0.0.0/0  |   TCP    | 2380 & 2379 | Etcd access |
+
+### Implementation Details/Notes/Constraints
+
+A new downstream OLM operator (openshift/ingress-nodefw-operator) will be designed to do the following:
+- Create Ingress node firewall config object to deploy daemonSet.
+- Create the CRDs for the new object.
+- controller for each of the new resources to handle add/update/delete
+- verification webhook to validate the configuration and enforce failSafe rules.
+- library to construct and load eBPF ingress maps tables based on the provisioned rules.
+- library for statistics collection and exporting to prometheus.
+- library for bpf perf event reporting.
+- library for FailSafe rules handling.
+- library to sync eBPF rules and update rules in eBPF maps.
+
+New downstream repo for the kernel hook implementation (openshift/ingress-node-firewall)
+- sample `IngressNodeConfig` object
+  
+```yaml
+apiVersion: ingressnodefirewall.openshift.io/v1alpha1
+kind: IngressNodeFirewallConfig
+metadata:
+  name: ingressnodefirewallconfig
+  namespace: ingress-node-firewall-system
+spec:
+  nodeSelector:
+    node-role.kubernetes.io/worker: ""
+  tolerations:
+  - key: "Example"
+    operator: "Exists"
+    effect: "NoExecute"
+```
+
+- sample `IngressNodeFirewall` object
+
+```yaml
+apiVersion: ingressnodefirewall.openshift.io/v1alpha1
+kind: IngressNodeFirewall
+metadata:
+  name: ingressnodefirewall-demo-1
+spec:
+  interfaces:
+  - eth0
+  nodeSelector:
+    matchLabels:
+      do-node-ingress-firewall: 'true'
+  ingress:
+  - sourceCIDRs:
+       - 1.1.1.1/24
+       - 100:1::1/64
+    rules:
+    - order: 10
+      protocolConfig:
+        protocol: TCP
+        tcp:
+          ports: "100-200"
+      action: Allow
+    - order: 20
+      protocolConfig:
+        protocol: UDP
+        udp:
+          ports: 8000
+      action: Allow
+  - sourceCIDRs:
+       - 3.3.3.3/24
+    rules:
+    - order: 10
+      protocolConfig:
+        protocol: ICMP
+        icmp:
+          icmpType: 3 # Destination unreachable
+          icmpCode: 1 # Host unreachable
+      action: Allow
+  - sourceCIDRs:
+       - 10:10::1/64
+    rules:
+    - order: 10
+      protocolConfig:
+        protocol: ICMPv6
+        icmpv6:
+          icmpType: 128 # Ping request
+      action: Deny
+  - sourceCIDRs:
+       - 0.0.0.0/0
+       - 0::0/0
+    rules:
+    - order: 10
+      protocolConfig:
+        protocol: TCP
+      action: Allow
+```
+
+In the case that there are multiple `IngressNodeFirewall` objects for the same nodes and interfaces with overlapping CIDRs, the `order` field will specify which rule applied first, lower order will be executed first and once the rule is executed and an action is taken, the firewall operation will be completed, and all the following rules won't be checked. 
```

Comment:
In the example above, all of the rules have `order: 10`... what does that mean? Is that even valid?

Reply:
different cidr can allow to use same order rule if u want within the same cidr the order has to be unique